### PR TITLE
Filter binaries by platform when EXTERNAL=true

### PR DESCRIPTION
**Pull request by cmacknz**
Not all binaries are available for every platform, and if we try to download a binary from an artifact URL that doesn't exist the build fails.

To reproduce the problem this solves, run `EXTERNAL=true SNAPSHOT=true PLATFORMS=darwin/arm64 PACKAGES=tar.gz mage -v package` on main. This is what I use to package the agent for my M1 Mac regularly and APM server not being published for arm64 causes the build to fail when fetching from the artifacts API.

This makes it so I don't have to manually remove APM server from the package list build every time I want build the agent.

### CODE REVIEW

**Review by oren-zohar — CHANGES_REQUESTED** [review]
Thanks for this fix @cmacknz. I think the issue here is rooted in the error handing in [e2e-testing](https://github.com/elastic/e2e-testing)'s `FetchBeatsBinary`, it shouldn't panic, and instead, it should propagate the error for outside handling. And on the agent side, whenever there's an error with fetching a specific artifact, I would suggest just printing it and moving forward with the build.

fortunately, I think this issue is solved with the latest `e2e-testing` package, this is what I did to verify:
1. Update package dependency to `github.com/elastic/e2e-testing v1.99.2-0.20221205111528-ade3c840d0c0`
2. Uncommented `cloudbeat` from the list of supported beats (I'll do it on my PR)
3. changed the panic here to just printing the error:   
  https://github.com/elastic/elastic-agent/blob/0e05b706399cfa380874ed1774315feb8440c475/magefile.go#LL668-L671C7

and these are the logs for running build command that is unsupported by both cloudbeat and APM
```bash
PLATFORMS=darwin/arm64 PACKAGES="docker" EXTERNAL=true SNAPSHOT=true mage -v package
```

Logs:
```bash
> PLATFORMS=darwin/arm64 PACKAGES="docker" EXTERNAL=true SNAPSHOT=true mage -v package
Running target: Package
WARN[0001] Download could not be found at the Elastic downloads API  elapsedTime=384.747708ms error="HEAD request failed with 404" resolver="{  false map[]    https://artifacts.elastic.co/downloads/beats/auditbeat/auditbeat-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz}" retry=1 statusEndpoint="https://artifacts.elastic.co/downloads/beats/auditbeat/auditbeat-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz"
WARN[0001] Object not found. Trying with another download resolver  resolver="&{beats auditbeat-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz auditbeat}"
downloaded binaries on location: /Users/orenzohar/GolandProjects/oren-zohar/forks/elastic-agent/build/distributions/elastic-agent-drop/archives/darwin-aarch64.tar.gz/auditbeat-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz.sha512
WARN[0014] Download could not be found at the Elastic downloads API  elapsedTime=211.779875ms error="HEAD request failed with 404" resolver="{  false map[]    https://artifacts.elastic.co/downloads/beats/filebeat/filebeat-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz}" retry=1 statusEndpoint="https://artifacts.elastic.co/downloads/beats/filebeat/filebeat-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz"
WARN[0014] Object not found. Trying with another download resolver  resolver="&{beats filebeat-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz filebeat}"
downloaded binaries on location: /Users/orenzohar/GolandProjects/oren-zohar/forks/elastic-agent/build/distributions/elastic-agent-drop/archives/darwin-aarch64.tar.gz/filebeat-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz.sha512
WARN[0027] Download could not be found at the Elastic downloads API  elapsedTime=202.77375ms error="HEAD request failed with 404" resolver="{  false map[]    https://artifacts.elastic.co/downloads/beats/heartbeat/heartbeat-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz}" retry=1 statusEndpoint="https://artifacts.elastic.co/downloads/beats/heartbeat/heartbeat-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz"
WARN[0027] Object not found. Trying with another download resolver  resolver="&{beats heartbeat-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz heartbeat}"
downloaded binaries on location: /Users/orenzohar/GolandProjects/oren-zohar/forks/elastic-agent/build/distributions/elastic-agent-drop/archives/darwin-aarch64.tar.gz/heartbeat-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz.sha512
WARN[0037] Download could not be found at the Elastic downloads API  elapsedTime=212.845042ms error="HEAD request failed with 404" resolver="{  false map[]    https://artifacts.elastic.co/downloads/beats/metricbeat/metricbeat-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz}" retry=1 statusEndpoint="https://artifacts.elastic.co/downloads/beats/metricbeat/metricbeat-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz"
WARN[0037] Object not found. Trying with another download resolver  resolver="&{beats metricbeat-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz metricbeat}"
downloaded binaries on location: /Users/orenzohar/GolandProjects/oren-zohar/forks/elastic-agent/build/distributions/elastic-agent-drop/archives/darwin-aarch64.tar.gz/metricbeat-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz.sha512
WARN[0052] Download could not be found at the Elastic downloads API  elapsedTime=217.831958ms error="HEAD request failed with 404" resolver="{  false map[]    https://artifacts.elastic.co/downloads/beats/osquerybeat/osquerybeat-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz}" retry=1 statusEndpoint="https://artifacts.elastic.co/downloads/beats/osquerybeat/osquerybeat-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz"
WARN[0052] Object not found. Trying with another download resolver  resolver="&{beats osquerybeat-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz osquerybeat}"
downloaded binaries on location: /Users/orenzohar/GolandProjects/oren-zohar/forks/elastic-agent/build/distributions/elastic-agent-drop/archives/darwin-aarch64.tar.gz/osquerybeat-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz.sha512
WARN[0068] Download could not be found at the Elastic downloads API  elapsedTime=196.64ms error="HEAD request failed with 404" resolver="{  false map[]    https://artifacts.elastic.co/downloads/beats/packetbeat/packetbeat-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz}" retry=1 statusEndpoint="https://artifacts.elastic.co/downloads/beats/packetbeat/packetbeat-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz"
WARN[0068] Object not found. Trying with another download resolver  resolver="&{beats packetbeat-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz packetbeat}"
downloaded binaries on location: /Users/orenzohar/GolandProjects/oren-zohar/forks/elastic-agent/build/distributions/elastic-agent-drop/archives/darwin-aarch64.tar.gz/packetbeat-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz.sha512
WARN[0079] Download could not be found at the Elastic downloads API  elapsedTime=210.837166ms error="HEAD request failed with 404" resolver="{  false map[]    https://artifacts.elastic.co/downloads/beats/apm-server/apm-server-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz}" retry=1 statusEndpoint="https://artifacts.elastic.co/downloads/beats/apm-server/apm-server-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz"
WARN[0079] Object not found. Trying with another download resolver  resolver="&{beats apm-server-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz apm-server}"
ERRO[0079] object not found in Artifact API              artifact=apm-server name=apm-server-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz version=8.7.0-SNAPSHOT
ERRO[0079] Object not found. There is no other download resolver 
downloaded binaries on location: 
WARN[0080] Download could not be found at the Elastic downloads API  elapsedTime=207.659334ms error="HEAD request failed with 404" resolver="{  false map[]    https://artifacts.elastic.co/downloads/beats/cloudbeat/cloudbeat-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz}" retry=1 statusEndpoint="https://artifacts.elastic.co/downloads/beats/cloudbeat/cloudbeat-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz"
WARN[0080] Object not found. Trying with another download resolver  resolver="&{beats cloudbeat-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz cloudbeat}"
ERRO[0080] object not found in Artifact API              artifact=cloudbeat name=cloudbeat-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz version=8.7.0-SNAPSHOT
ERRO[0080] Object not found. There is no other download resolver 
downloaded binaries on location: 
WARN[0080] Download could not be found at the Elastic downloads API  elapsedTime=201.548834ms error="HEAD request failed with 404" resolver="{  false map[]    https://artifacts.elastic.co/downloads/beats/elastic-agent-shipper/elastic-agent-shipper-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz}" retry=1 statusEndpoint="https://artifacts.elastic.co/downloads/beats/elastic-agent-shipper/elastic-agent-shipper-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz"
WARN[0080] Object not found. Trying with another download resolver  resolver="&{beats elastic-agent-shipper-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz elastic-agent-shipper}"
downloaded binaries on location: /Users/orenzohar/GolandProjects/oren-zohar/forks/elastic-agent/build/distributions/elastic-agent-drop/archives/darwin-aarch64.tar.gz/elastic-agent-shipper-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz.sha512
WARN[0084] Download could not be found at the Elastic downloads API  elapsedTime=203.390375ms error="HEAD request failed with 404" resolver="{  false map[]    https://artifacts.elastic.co/downloads/beats/endpoint-security/endpoint-security-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz}" retry=1 statusEndpoint="https://artifacts.elastic.co/downloads/beats/endpoint-security/endpoint-security-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz"
WARN[0084] Object not found. Trying with another download resolver  resolver="&{beats endpoint-security-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz endpoint-security}"
downloaded binaries on location: /Users/orenzohar/GolandProjects/oren-zohar/forks/elastic-agent/build/distributions/elastic-agent-drop/archives/darwin-aarch64.tar.gz/endpoint-security-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz.sha512
WARN[0095] Download could not be found at the Elastic downloads API  elapsedTime=206.781167ms error="HEAD request failed with 404" resolver="{  false map[]    https://artifacts.elastic.co/downloads/beats/fleet-server/fleet-server-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz}" retry=1 statusEndpoint="https://artifacts.elastic.co/downloads/beats/fleet-server/fleet-server-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz"
WARN[0095] Object not found. Trying with another download resolver  resolver="&{beats fleet-server-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz fleet-server}"
downloaded binaries on location: /Users/orenzohar/GolandProjects/oren-zohar/forks/elastic-agent/build/distributions/elastic-agent-drop/archives/darwin-aarch64.tar.gz/fleet-server-8.7.0-SNAPSHOT-darwin-aarch64.tar.gz.sha512
exec: go "list" "-m"
Found Elastic Beats dir at /Users/orenzohar/GolandProjects/oren-zohar/forks/elastic-agent
elastic_beat_agent_binaries package spec loaded from [/Users/orenzohar/GolandProjects/oren-zohar/forks/elastic-agent/dev-tools/packaging/packages.yml]
Running dependency: Update
Running dependency: Config
Running dependency: configYML
>> Building elastic-agent.yml for linux/amd64
>> Building elastic-agent.reference.yml for linux/amd64
>> Building elastic-agent.docker.yml for linux/amd64
Running dependency: BuildPGP
>> BuildPGP from GPG-KEY-elasticsearch to internal/pkg/release/pgp.go
exec: go "run" "dev-tools/cmd/buildpgp/build_pgp.go" "--in" "GPG-KEY-elasticsearch" "--out" "internal/pkg/release/pgp.go"
Running dependency: BuildFleetCfg
>> BuildFleetCfg _meta/elastic-agent.fleet.yml to internal/pkg/agent/application/configuration_embed.go
exec: go "run" "dev-tools/cmd/buildfleetcfg/buildfleetcfg.go" "--in" "_meta/elastic-agent.fleet.yml" "--out" "internal/pkg/agent/application/configuration_embed.go"
Running dependency: CrossBuildGoDaemon
Skipping cross-build of target=buildGoDaemon because platforms list is empty.
Running dependency: CrossBuild
exec: docker "info" "-f" "{{ json .}}"
Running dependency: github.com/elastic/elastic-agent/dev-tools/mage.CrossBuild.func1
exec: go "mod" "download"
Running dependency: github.com/elastic/elastic-agent/dev-tools/mage.buildMage
exec: mage "-f" "-goos=linux" "-goarch=arm64" "-compile" "build/mage-linux-arm64"
...
```

**Comment by cmacknz** [comment]
Thanks @oren-zohar, didn't think to check that this was fixed in the e2e-testing package. That simplifies things nicely, although there are still some changes to the e2e-testing package we could follow up with. Setting the log level could be easier, and a defined error type for NotFound would be nice.

File: go.mod [hunk 1]
```diff
@@ -11,12 +11,14 @@ require (
 	github.com/coreos/go-systemd/v22 v22.3.3-0.20220203105225-a9a7ef127534
 	github.com/docker/go-units v0.4.0
 	github.com/dolmen-go/contextio v0.0.0-20200217195037-68fc5150bcd5
-	github.com/elastic/e2e-testing v1.99.2-0.20220117192005-d3365c99b9c4
+	github.com/elastic/beats/v7 v7.17.7
```
Comment (cmacknz) [inline go.mod:14]: Ugh, I didn't notice this come in. We have been trying to keep Beats out of the agent dependency list. I'll see if I can get rid of this.

File: magefile.go [hunk 2]
```diff
@@ -902,8 +912,11 @@ func fetchBinaryFromArtifactsApi(ctx context.Context, packageName, artifact, ver
 		false,
 		downloadPath,
 		true)
-	fmt.Println("downloaded binaries on location:", location)
+	if err != nil {
+		return err
+	}
 
+	fmt.Println("downloaded binaries on location:", location)
```
Comment (AndersonQ) [inline magefile.go:919]: [Suggestion]
It feels a bit redundant
```suggestion
	fmt.Println("downloaded binaries on", location)
```

File: magefile.go [hunk 3]
```diff
@@ -650,24 +647,29 @@ func packageAgent(requiredPackages []string, packagingFn func()) {
...
 					targetPath := filepath.Join(archivePath, reqPackage)
 					os.MkdirAll(targetPath, 0755)
-					newVersion, packageName := getPackageName(beat, version, reqPackage)
-					err := fetchBinaryFromArtifactsApi(ctx, packageName, beat, newVersion, targetPath)
+					newVersion, packageName := getPackageName(binary, version, reqPackage)
+					err := fetchBinaryFromArtifactsApi(ctx, packageName, binary, newVersion, targetPath)
 					if err != nil {
-						panic(fmt.Sprintf("fetchBinaryFromArtifactsApi failed: %v", err))
+						if strings.Contains(err.Error(), "object not found") {
+							fmt.Printf("Downloading %s: unsupported on %s, skipping\n", binary, platform)
```
Comment (AndersonQ) [inline magefile.go:669]: It seems to me too fragile just rely on the fact that the object isn't there to be sure the platform is unsupported. 

I believe the best would be the agent only try to download the supported binaries, and fail the build if they cannot be downloaded.

**Review by AndersonQ — CHANGES_REQUESTED** [review]
https://github.com/elastic/elastic-agent/pull/1908#discussion_r1046243388 isn't exacly a blocker, but I'd like to see what you think @cmacknz

File: magefile.go (hunk 3 above)
Comment (cmacknz) [reply magefile.go:669]: This is hitting the official Elastic artifacts API, so if it isn't there it is a fairly strong indicator that the binary isn't supported. My first attempt at this was actually to encode which binaries were supported where into this magefile, but that is less likely to stay up to date than just trying to download the officially published artifacts themselves.

This path is only hit when building locally, it isn't actually used in the final release process which manually downloads each dependency from the DRA staging buckets. There's low risk of this causing us problems in the future.

File: go.mod (hunk 1 above)
Comment (cmacknz) [reply go.mod]: Had to clear out my local build/package directory.

File: magefile.go (hunk 3 above)
Comment (AndersonQ) [reply magefile.go:669]: good point, I'm happy with this solution :)

**Review by AndersonQ — APPROVED** [review]

**Comment by cmacknz** [comment]
@oren-zohar after the latest update are there any changes you'd still like made to this PR?

**Comment by oren-zohar** [comment]
@cmacknz let me recheck this and I'll update you

**Comment by oren-zohar** [comment]
Everything works as expected with cloudbeat, thanks @cmacknz 

**How I tested it**
Checked out your branch and added `cloudbeat` to the suppreted `EXTERNAL` beats and tested two commands:

```zsh
SNAPSHOT=true DEV=true EXTERNAL=true PLATFORMS=linux/amd64 TYPES=docker mage package -v
```
this platform is supported by cloudbeat and everything worked as expected

results with `windows/amd64` which is not supported by cloudbeat:

```zsh
> SNAPSHOT=true DEV=true EXTERNAL=true PLATFORMS=windows/amd64 TYPES=docker mage package -v
downloaded binaries on /Users/orenzohar/GolandProjects/oren-zohar/forks/elastic-agent/build/distributions/elastic-agent-drop/archives/windows-x86_64.zip/auditbeat-8.7.0-SNAPSHOT-windows-x86_64.zip.sha512
downloaded binaries on /Users/orenzohar/GolandProjects/oren-zohar/forks/elastic-agent/build/distributions/elastic-agent-drop/archives/windows-x86_64.zip/filebeat-8.7.0-SNAPSHOT-windows-x86_64.zip.sha512
downloaded binaries on /Users/orenzohar/GolandProjects/oren-zohar/forks/elastic-agent/build/distributions/elastic-agent-drop/archives/windows-x86_64.zip/heartbeat-8.7.0-SNAPSHOT-windows-x86_64.zip.sha512
downloaded binaries on /Users/orenzohar/GolandProjects/oren-zohar/forks/elastic-agent/build/distributions/elastic-agent-drop/archives/windows-x86_64.zip/metricbeat-8.7.0-SNAPSHOT-windows-x86_64.zip.sha512
downloaded binaries on /Users/orenzohar/GolandProjects/oren-zohar/forks/elastic-agent/build/distributions/elastic-agent-drop/archives/windows-x86_64.zip/osquerybeat-8.7.0-SNAPSHOT-windows-x86_64.zip.sha512
downloaded binaries on /Users/orenzohar/GolandProjects/oren-zohar/forks/elastic-agent/build/distributions/elastic-agent-drop/archives/windows-x86_64.zip/packetbeat-8.7.0-SNAPSHOT-windows-x86_64.zip.sha512
Downloading cloudbeat: unsupported on windows/amd64, skipping
...
```

**Review by oren-zohar — APPROVED** [review]

**Pull request merged**